components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java

```diff
@@ -0,0 +1,260 @@
+ * Implementation for {@link ServiceURLBuilder}.
+ * Builder for {@link ServiceURL} instances.
+ */
+public class DefaultServiceURLBuilder implements ServiceURLBuilder {
+
+    private boolean addProxyContext;
+    private boolean addWebContext;
+    private String urlContext;
+    private boolean appendTenantAsQueryParam;
+    private boolean appendTenantAsPathParam;
+    private Map<String, String> parameters;
+    private Map<String, String> fragments;
+
+    /**
+     * This method is called to add the URL context to the builder.
+     *
+     * @param context URL context.
+     * @return {@link ServiceURLBuilder}.
+     */
+    @Override
+    public ServiceURLBuilder addContext(String context) {
+
+        this.urlContext = context;
+        return this;
+    }
+
+    /**
+     * This method is called to add the proxy context path to the URL.
+     *
+     * @return {@link ServiceURLBuilder}.
+     */
+    @Override
+    public final ServiceURLBuilder addProxyContext() {
+
+        this.addProxyContext = true;
+        return this;
+    }
+
+    /**
+     * This method is called to add the web context root to the URL.
+     *
+     * @return {@link ServiceURLBuilder}.
+     */
+    @Override
+    public final ServiceURLBuilder addWebContext() {
+
+        this.addWebContext = true;
+        return this;
+    }
+
+    /**
+     * This method is called when the URL needs to be appended with the tenant domain as a query parameter when
+     * operating in legacy mode.
+     *
+     * @return {@link ServiceURLBuilder}.
+     */
+    @Override
+    public ServiceURLBuilder appendTenantAsQueryParam() {
+
+        this.appendTenantAsQueryParam = true;
+        return this;
+    }
+
+    /**
+     * This method is called when the URL needs to be appended with the tenant domain as a path parameter when
+     * operating in legacy mode.
+     *
+     * @return {@link ServiceURLBuilder}.
+     */
+    @Override
+    public ServiceURLBuilder appendTenantAsPathParam() {
+
+        this.appendTenantAsPathParam = true;
+        return this;
+    }
+
+    /**
+     * This method is called when the URL needs to be appended with other parameters. Such parameters should be
+     * entered as <k,v> pairs.
+     *
+     * @param key   Key.
+     * @param value Value.
+     * @return {@link ServiceURLBuilder}.
+     */
+    @Override
+    public ServiceURLBuilder addParameter(String key, String value) {
+
+        this.parameters.put(key, value);
+        return this;
+    }
+
+    /**
+     * This method is called when the URL needs to be appended with fragments. Fragments should be
+     * entered as <k,v> pairs.
+     *
+     * @param key   Key.
+     * @param value Value.
+     * @return {@link ServiceURLBuilder}.
+     */
+    @Override
+    public ServiceURLBuilder addFragments(String key, String value) {
+
+        this.fragments.put(key, value);
+        return this;
+    }
+
+    @Override
+    public ServiceURL build() throws URLBuilderException {
+
```

Comment:
The very first check we should do here if if tenant url mode is enabled. If not there's a config matched to the context we should return it.
Else we should build